File: web_cabinet/ai/prompts/morning_brief.py
"""Промпт для утреннего брифинга фермы (MVP-N14)."""
from __future__ import annotations

import json
from typing import Any

MORNING_BRIEF_SYSTEM = """\
Ты — ИИ-аналитик GenomeAI. Каждое утро ты готовишь структурированный \
оперативный брифинг для руководителя молочной фермы.

ЯЗЫК: Строго русский. Профессиональная лексика молочного животноводства.

ПРАВИЛА:
1. Каждый факт о конкретном животном или показателе ОБЯЗАТЕЛЬНО сопровождается \
evidence_id из контекста (event_id из health_events, treatments, repro_events).
2. Утверждения без evidence_id — НЕ включать.
3. ЗАПРЕЩЕНО: обобщения без данных, «обычно», «как правило», «вероятно» без цифр, \
придуманные данные.
4. Объём main_takeaway: 2–3 предложения, только самое важное.

HEADLINE (однострочный заголовок, выбери один):
- "Спокойное утро — плановая работа" (нет критических событий за ночь)
- "Требуется внимание: [конкретная проблема]" (есть важные события)
- "Критично: [краткое описание]" (критические события, требуют немедленных действий)

OVERNIGHT_CHANGES: события/изменения за последние 24 часа с evidence_id.
TODAY_ACTIONS: конкретные действия на сегодня (до 5), с приоритетом и ответственным:
  priority: "high" | "medium" | "low"
  role: "vet" | "zootech" | "operator" | "director"
  due: время до которого выполнить ("до 10:00", "до 14:00") или null

Верни ТОЛЬКО валидный JSON (без markdown-обёртки):\
"""

_RESPONSE_SCHEMA = """\
{
  "headline": "<однострочный заголовок>",
  "main_takeaway": "<2-3 предложения — самое важное>",
  "overnight_changes": [
    {"text": "<факт с цифрами>", "evidence_id": "<event_id или null>"}
  ],
  "today_actions": [
    {"action": "<конкретное действие>", "priority": "high", "due": "<до HH:MM или null>", "role": "vet"}
  ],
  "notes": ["<дополнительное наблюдение, необязательно>"]
}\
"""


def build_morning_brief_message(context: Any, date: str) -> str:
    """Строит user message для утреннего брифинга."""
    if hasattr(context, "to_text"):
        context_text = context.to_text()
    elif isinstance(context, dict):
        context_text = json.dumps(context, ensure_ascii=False, default=str, indent=2)
    else:
        context_text = str(context)

    return (
        f"Дата брифинга: {date}\n\n"
        f"<farm_context>\n{context_text}\n</farm_context>\n\n"
        f"Подготовь утренний брифинг для руководителя фермы на {date}.\n"
        f"Верни результат строго в формате JSON (без обёртки ```json):\n\n"
        f"{_RESPONSE_SCHEMA}"
    )

File: web_app/lib/api/morning-brief.ts
import { apiFetch } from '@/lib/api/client';

export interface OvernightChange {
  text: string;
  evidence_id: string | null;
}

export interface TodayAction {
  action: string;
  priority: 'high' | 'medium' | 'low';
  due: string | null;
  role: 'vet' | 'zootech' | 'operator' | 'director';
}

export interface MorningBrief {
  brief_id: string;
  farm_id: string;
  generated_at_utc: string;
  date: string;
  headline: string;
  main_takeaway: string;
  overnight_changes: OvernightChange[];
  today_actions: TodayAction[];
  notes: string[];
  generation_model: string;
  generation_tokens: { input: number; output: number };
}

export async function fetchMorningBrief(farmId = 'INV_FARM_001'): Promise<MorningBrief> {
  return apiFetch<MorningBrief>(`/api/ai/morning-brief/today?farm_id=${encodeURIComponent(farmId)}`);
}

export async function regenerateMorningBrief(farmId = 'INV_FARM_001'): Promise<MorningBrief> {
  return apiFetch<MorningBrief>('/api/ai/morning-brief', {
    method: 'POST',
    body: JSON.stringify({ farm_id: farmId, force_regenerate: true }),
  });
}

export function morningBriefPdfUrl(briefId: string, farmId = 'INV_FARM_001'): string {
  return `/api/backend/api/ai/morning-brief/${briefId}/pdf?farm_id=${encodeURIComponent(farmId)}`;
}

export interface ApproveBriefResult {
  approved: boolean;
  tasks_created: number;
}

export async function approveMorningBrief(
  briefId: string,
  actions: TodayAction[],
  farmId = 'INV_FARM_001',
): Promise<ApproveBriefResult> {
  return apiFetch<ApproveBriefResult>(`/api/ai/morning-brief/${encodeURIComponent(briefId)}/approve`, {
    method: 'POST',
    body: JSON.stringify({ farm_id: farmId, actions }),
  });
}
